### ISSUE TITLE
Cannot create a mask for each sequence in a batch with Flex Attention

### ISSUE THREAD

**Issue by luispl77** [pytorch/pytorch#157675]
### 🚀 The feature, motivation and pitch

Hello everyone,

I have been experimenting with Flex Attention when I ran into a problem that makes Flex Attention useless for my use case:

```python
def causal_mask(b, h, q_idx, kv_idx):
    return q_idx >= kv_idx
```
I require applying a **custom attention mask** (not the causal one above) **that is computed based on the specific sequence of tokens themselves** (that is how my custom architecture works). So in a batch of 4 sequences, I might have 4 different masks, each depending on the sequence itself. I compute the masks outside this function, and store them in a tensor, and because this mask_mod function can read any tensor on the device, I can have a custom mask based on the sequence.
 But...

I initially assumed that "b" is the batch index, but its not. 
Without knowing which batch I am in, I cannot compute my own mask for each sequence. I am left with applying the same mask for the entire batch, and thus I cannot use Flex Attention to implement my architecture at scale, that is, using vLLM, with large batches.

I got quite far in patching the vLLM code and using the newly added Flex Attention backend. But I run into that same issue. vLLM works fenominaly well on batches of sequences. And because of the flaw with Flex Attention (not being able to set a mask for each sequence), I am restricted to maybe doing one sequence at a time in vLLM. 

Could someone help me confirm that this indeed is a limitation that exists, and if so, will there be plans to fix this?
Because I think if indeed I am correct, it makes Flex Attention not very useful for a wide range of architectures. We want to be able to have a custom attention mask, but not just "blind" attention patterns like checkerboard or sliding windows. Those patterns seldom make any difference. What is much more interesting are patterns **based on the content of the sequence themselves**, which is what I am trying to achieve. 

Thank you!

cc @chauhang @penguinwu @zou3519 @ydwu4 @bdhirsh @Chillee @drisspg @yanboliang @BoyuanFeng

**Comment by Chillee** [comment]
Generally speaking `b` is the batch-index. In VLLM there are some additional challenges due to continuous batching - I'm not quite sure they'll work yet in vllm.

**Comment by luispl77** [comment]
Ok, this is good that b is the batch index.
I wasn't sure because I can't print it, so I was trying to use conditions to change the mask like b == 0 or b == 1, and in these tests, it seems that b was always 0. Therefore, the problem is might not be with Flex Attention but with the fact that perhaps vLLM calls Flex Attention for each sequence individually, and that is why i am only getting b == 0 on every sequence in the batch, rather than an index for each.
So the issue might even be with vLLM then. I will have to continue investigating over there. 
Thanks anyway

Edit: also the documentation says "b" is the batch size, might want to take a look at that https://docs.pytorch.org/docs/stable/nn.attention.flex_attention.html#blockmask-utilities

**Comment by drisspg** [comment]
B is always the the batch index, if the input is 4d. However there are cases where people rearrange the input into a "stack of ragged sequences" an have a technical batch size of 1. This is the case in vllm. A simpler example that illustrates this does "document masking":  https://github.com/pytorch-labs/attention-gym/blob/main/attn_gym/masks/document_mask.py
